plugins/traefik/main.go

```diff
@@ -136,11 +164,132 @@ func (r *responseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	if !ok {
 		return nil, nil, fmt.Errorf("%T is not a http.Hijacker", r.responseWriter)
 	}
-	return hijacker.Hijack()
+	if r.websocket {
+		fmt.Println("=== hijack for websocket")
+		conn, bufio, err := hijacker.Hijack()
+		return newConnWrapper(conn), bufio, err
+	} else {
+		return hijacker.Hijack()
+	}
 }
 
 func (r *responseWriter) Flush() {
 	if flusher, ok := r.responseWriter.(http.Flusher); ok {
 		flusher.Flush()
 	}
 }
+
+func isWebsocketRequest(req *http.Request) bool {
+	return containsHeader(req, "Connection", "upgrade") && containsHeader(req, "Upgrade", "websocket")
+}
+
+func containsHeader(req *http.Request, name, value string) bool {
+	items := strings.Split(req.Header.Get(name), ",")
+	for _, item := range items {
+		if value == strings.ToLower(strings.TrimSpace(item)) {
+			return true
+		}
+	}
+	return false
+}
+
+func newConnWrapper(c net.Conn) *conn {
+	return &conn{
+		conn: c,
+	}
+}
+
+type conn struct {
+	conn net.Conn
+}
+
+// LocalAddr implements net.Conn.
+func (c *conn) LocalAddr() net.Addr {
+	panic("unimplemented")
+}
+
+// RemoteAddr implements net.Conn.
+func (c *conn) RemoteAddr() net.Addr {
+	panic("unimplemented")
+}
+
+// SetDeadline implements net.Conn.
+func (c *conn) SetDeadline(t time.Time) error {
+	panic("unimplemented")
+}
+
+// SetReadDeadline implements net.Conn.
+func (c *conn) SetReadDeadline(t time.Time) error {
+	panic("unimplemented")
+}
+
+// SetWriteDeadline implements net.Conn.
+func (c *conn) SetWriteDeadline(t time.Time) error {
+	panic("unimplemented")
+}
+
+func (c *conn) Read(b []byte) (n int, err error) {
+	n, err = c.conn.Read(b)
+	if err == nil {
+		wsEventChan <- WebSocketRead // Notify about the read operation
+	}
+	return
+}
+
+func (c *conn) Write(b []byte) (n int, err error) {
+	n, err = c.conn.Write(b)
+	if err == nil {
+		wsEventChan <- WebSocketWrite // Notify about the write operation
+	}
+	return
+}
+
+func (c *conn) Close() error {
+	err := c.conn.Close()
+	wsEventChan <- WebSocketClose // Notify about the close operation
+	return err
+}
+
+func monitorWebSocketActivity(sablierRequest *http.Request, sm *SablierMiddleware) {
+	duration, err := time.ParseDuration(sm.config.SessionDuration)
+	if err != nil {
+		fmt.Fprintln(os.Stdout, []any{`Error parsing sessionDuration: %v`, err}...)
+		return
+	}
+	alertTime := duration - (duration * 5 / 100) // Calculate alert time at 95% of the total duration
```

Comment:
I would do just half of the value which is simpler and more reliable imho.

Reply:
i do this, 3-5 % of the duration , because, i need to trig Sablier but not on every  `event`, with this mechanism, i can refresh the scaling up( if i don't the Sablier scale down) with http

and i do not close on `WebSQocketClose` directly because u still have to wait the session duration

Reply:
but the half duration make sense to me too

Reply:
Now I get your point. It means that timeout will have random value depending on when the last request was done. It will be a little confusing. It would be better to make another request when connection is closed and cancel the timer. That way the timeout would be predictable from user point of view.